libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,655 @@
+#  else
+template <class _Tp, class... _Tail>
+struct __extract_last_impl : __extract_last_impl<_Tail...> {};
+template <class _Tp>
+struct __extract_last_impl<_Tp> {
+  using type _LIBCPP_NODEBUG = _Tp;
+};
+
+template <class... _Tp>
+using __extract_last _LIBCPP_NODEBUG = __extract_last_impl<_Tp...>::type;
+#  endif
+
+template <bool _Const, class... _Tp>
+struct __all_but_first_model_sized_range;
+
+template <bool _Const, class _Head, class... _Tail>
+struct __all_but_first_model_sized_range<_Const, _Head, _Tail...> {
+  static constexpr bool value = (sized_range<__maybe_const<_Const, _Tail>> && ...);
+};
+
+template <bool _Const, class... _Views>
+concept __all_random_access = (random_access_range<__maybe_const<_Const, _Views>> && ...);
+
+template <bool _Const, class... _Views>
+concept __all_bidirectional = (bidirectional_range<__maybe_const<_Const, _Views>> && ...);
+
+template <bool _Const, class... _Views>
+concept __all_forward = (forward_range<__maybe_const<_Const, _Views>> && ...);
+
+template <bool _Const, class _First, class... _Tail>
+struct __all_common_ignore_last {
+  static constexpr bool value =
+      common_range<__maybe_const<_Const, _First>> && __all_common_ignore_last<_Const, _Tail...>::value;
+};
+
+template <bool _Const, class _Tail>
+struct __all_common_ignore_last<_Const, _Tail> {
+  static constexpr bool value = true;
+};
+
+template <bool _Const, class... _Rs>
+concept __concat_is_random_access =
+    (__all_random_access<_Const, _Rs...>) && (__all_common_ignore_last<_Const, _Rs...>::value);
+
+template <bool _Const, class... _Rs>
+concept __concat_is_bidirectional =
+    (__all_bidirectional<_Const, _Rs...>) && (__all_common_ignore_last<_Const, _Rs...>::value);
+
+template <input_range... _Views>
+  requires((view<_Views> && ...) && (sizeof...(_Views) > 0) && __concatable<_Views...>)
+class concat_view : public view_interface<concat_view<_Views...>> {
+  tuple<_Views...> __views_;
+
+  template <bool _Const>
+  class __iterator;
+
+public:
+  _LIBCPP_HIDE_FROM_ABI constexpr concat_view() = default;
+
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit concat_view(_Views... __views) : __views_(std::move(__views)...) {}
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator<false> begin()
+    requires(!(__simple_view<_Views> && ...))
+  {
+    __iterator<false> __it(this, in_place_index<0>, ranges::begin(std::get<0>(__views_)));
+    __it.template __satisfy<0>();
+    return __it;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator<true> begin() const
+    requires((range<const _Views> && ...) && __concatable<const _Views...>)
+  {
+    __iterator<true> __it(this, in_place_index<0>, ranges::begin(std::get<0>(__views_)));
+    __it.template __satisfy<0>();
+    return __it;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr auto end()
+    requires(!(__simple_view<_Views> && ...))
+  {
+    if constexpr (__all_forward<false, _Views...> && common_range<__maybe_const<false, __extract_last<_Views...>>>) {
```

Comment:
thanks for applying the fix for https://cplusplus.github.io/LWG/issue4166
BTW, it is tentatively ready now and will likely to be merged in WD in two weeks time.

could you please add test coverage for this case? (the first range is an input only range, e.g istream_view,  and the last range is a common range)